### ISSUE TITLE
DLC Queuing Produces Install Error

### ISSUE THREAD

**Issue by acook** [sharkwouter/minigalaxy#579]
When I clicked multiple DLCs to install, *all* will fail with errors either immediately or after the download completes. In addition the download bar will jump rapidly around.

There are two errors that pop up: "Failed to install" DLC and DLC "Corrupted"

![Screenshot from 2023-08-26 09-44-53](https://github.com/sharkwouter/minigalaxy/assets/71984/44b425ad-f8e3-4f0c-b5d8-17030a8f540f)

Original DLC queuing PR: https://github.com/sharkwouter/minigalaxy/pull/395

**Comment by sharkwouter** [comment]
Thanks for reporting this. I tried to install DLC for AI War to see if I could reproduce this, but there I wasn't seeing this issue. Could you run Minigalaxy from the command line and see what the output is when you try this again? Also, which version of Minigalaxy is this?

**Comment by acook** [comment]
Version `1.2.5`. I also saw the same issue with Battletech.

I'll see if I can replicate it with a non-Paradox game as well.

**Comment by NonMeme** [comment]
Just letting you know, I had an error installing a Stellaris DLC pack last year and a new one now where I cannot get the whole game to install. This one seems to be kind of finicky with MG.

Your error messages are completely different than mine, so I cannot quite help other than saying that.

**Comment by sharkwouter** [comment]
Thanks for updating @NonMeme, that's appreciated. What error message do you get?

**Comment by NonMeme** [comment]
> Thanks for updating @NonMeme, that's appreciated. What error message do you get?

[These.](https://github.com/sharkwouter/minigalaxy/issues/586) And I've tried to reinstall the game again off and on this month and last year too. It says on the Minigalaxy menu "in queue" and never goes to downloading.

**Comment by GB609** [comment]
About the jumping around of the progress bar: I've had the same. This happens when multiple large files are downloaded for a game, e.g. for some more recent windows games.
Cause:
Downloads for all files are started in parallel and download_manager reports progress for each file individually whenever a chunk is downloaded.
A percentage is calculated for the file which somehow tries to also consider that there are multiple files: file% * filenum / filecount. When there are 3 files, the first one thus is mapped 0-30, the second 30-60 etc. This value  is simply passed through to gametile.set_progress. This sets the actual bar. This could work in theory, if the downloads were sequential. But they are not run.

I see 2 ways to fix this:
1. Make downloads sequential
2. Let each download report progress in bytes and calculate the total percentage in gametile.set_progress which must keep track of current and total size per file.

**Comment by sharkwouter** [comment]
This could be related to the wine in the prefix not being used to install the dlc, but the system's wine instead. That is if you do not have Innoextract installed anyway.

**Comment by GB609** [comment]
I think the original error showed an .sh script, meanings its a linux game.
But i also get a lot of corrupted downloads in general. Could there be an issue with how downloaded chunks are consumed?

**Comment by sharkwouter** [comment]
Good point.

Yeah, the corrupted downloads are a separate issue. Sometimes resuming a download after restarting Minigalaxy corrupts it. I think I can fix that.

**Comment by GB609** [comment]
Small update about the download corruption that happened for me:
It appears i had a faulty combination of BIOS Firmware and RAM which also got me mayor issues when upgrading my system with pacman.
Since I upgraded the BIOS of my gaming rig, the download corruption has mostly disappeared.

**Comment by GB609** [comment]
About the initial error: I've just encountered something similar while testing my latest changes.

The problem is that the `__download` function empties the installer director of the game unconditionally everytime it is used. That function is invoked for the regular game download and for each click on download dlc individually.
At the end, only the last file (DLC) downloaded will remain. The download still works because thats just how the linux file system behaves: Deleting a file that is open will keep the file visible for all processes still holding the handle, but not for new ones. After the download is finished, minigalaxy tries to spawn a sub-process for the file and this is where it will fail, because the file will then be gone already.

This is related to #513 

Coupled with the preference 'Keep installers' this leads to a state in which even uninstalling and reinstalling the game does not fix anything. I've had this with Fell Seal.
Because minigalaxy appears to only check for existance of an executable in the game's installer directory (to skip re-downloading) and tries to run that.
So the cycle:
1. download game
2. install game
3. download DLC
4. install DLC
5. uninstall game

will end up with the game uninstalled, the base installer deleted and only the DLC installer remaining.

When trying to install from there, minigalaxy thinks that the DLC installer is the game's base installer and runs that.
In my case, the install itself was a success, but the attempt to create the desktop shortcut afterwards failed (because the DLC's gog*.info file does not contain playTasks and the DLC only directory didnt have any executable):
```text
warning [/home/.../GOG Games/installer/Fell Seal Arbiters Mark/fell_seal_arbiter_s_mark_missions_and_monsters_1_6_0_49888.sh]:  870358 extra bytes at beginning or within zipfile
  (attempting to process anyway)
2025-02-16 11:22:50,545 - minigalaxy - ERROR - Error installing game Fell Seal: Arbiter's Mark
Traceback (most recent call last):
  File "/home/.../git/minigalaxy/minigalaxy/installer.py", line 75, in install_game
    error_message = create_applications_file(game)
  File "/home/.../git/minigalaxy/minigalaxy/installer.py", line 281, in create_applications_file
    "game_bin_path": get_exec_line(game),
                     ~~~~~~~~~~~~~^^^^^^
  File "/home/.../git/minigalaxy/minigalaxy/installer.py", line 257, in get_exec_line
    exe_cmd_list = get_execute_command(game)
  File "/home/.../git/minigalaxy/minigalaxy/launcher.py", line 77, in get_execute_command
    exe_cmd = get_final_resort_exe_cmd(game, files)
  File "/home/.../git/minigalaxy/minigalaxy/launcher.py", line 200, in get_final_resort_exe_cmd
    exe_cmd = ["./{}".format(info["playTasks"][0]["path"])]

```

The only remedy for that would be to either manually remove the installer directory or disable the preference 'keep installers'.
